File: problems/calculate_amount_paid_in_taxes/calculate_amount_paid_in_taxes.cc
#include "calculate_amount_paid_in_taxes.h"

#include <algorithm>
#include <vector>

using std::min;
using std::vector;

double calculateAmountPaidInTaxes(const vector<vector<int>>& brackets, const int income) {
    double totalTax = 0.0;

    int incomeRemaining = income;
    int previousUpperLimit = 0;

    for (const auto& bracket : brackets) {
        const int upperLimit = bracket[0];
        const int ratePercent = bracket[1];

        // If no income remains, break early.
        if (incomeRemaining <= 0) {
            return totalTax;
        }

        // Calculate the taxable amount in the current bracket.
        const int bracketSize = upperLimit - previousUpperLimit;
        const int taxableAmount = min(incomeRemaining, bracketSize);

        // Add the tax for this bracket.
        totalTax += taxableAmount * (ratePercent / 100.0);

        // Update remaining income and previous upper limit.
        incomeRemaining -= taxableAmount;
        previousUpperLimit = upperLimit;
    }

    return totalTax;
}

File: problems/calculate_amount_paid_in_taxes/calculate_amount_paid_in_taxes.py
from typing import List


def calculateAmountPaidInTaxes(brackets: List[List[int]], income: int) -> float:
    total_tax = 0.0

    income_remaining = income
    previous_upper_limit = 0

    for upper_limit, rate_percent in brackets:
        # If no income remains, break early.
        if income_remaining <= 0:
            return total_tax

        # Calculate the taxable amount in the current bracket.
        bracket_size = upper_limit - previous_upper_limit
        taxable_amount = min(income_remaining, bracket_size)

        # Add the tax for this bracket.
        total_tax += taxable_amount * (rate_percent / 100.0)

        # Update remaining income and previous upper limit.
        income_remaining -= taxable_amount
        previous_upper_limit = upper_limit

    return total_tax
